docs/typescript/testing.md

```diff
@@ -2,92 +2,246 @@
+
+The built-in test server automatically "skips" (fast forwards) time when no activities are executing.
+If a Workflow sleeps for days, running it in the test environment will cause it to complete almost immediately.
 
-  it('bubbles up activity errors', async function () {
-    sinon
-      .stub(axios, 'get')
-      .callsFake(() => Promise.reject(new Error('example error')));
-
-    const err = await client.execute(example, { taskQueue: 'testhttp' }).then(
-      () => null,
-      (err) => err
-    );
-    assert.equal(err.name, 'WorkflowFailedError');
-    assert.equal(err.cause.cause.message, 'example error');
+<details>
+<summary>
+Workflow implementation
+</summary>
+
+<!--SNIPSTART typescript-timer-reminder-workflow-->
+<!--SNIPEND-->
+
+</details>
+
+```ts
+test('countdownWorkflow sends reminder email if processing does not complete in time', async () => {
+  // NOTE: this tests doesn't actually take days to complete, the test environment starts a test
+  // server that automatically skips time when there are no running activities.
+  let emailSent = false;
+  // createActivities defintion omitted for brevity
+  const activities: ReturnType<typeof createActivities> = {
+    async processOrder() {
+      // Test server switches to "normal" time while an activity is executing.
+      // Call `sleep` to skip time by "2 days".
+      await testEnv.sleep('2 days');
+    },
+    async sendNotificationEmail() {
+      emailSent = true;
+    },
+  };
+  const worker = await Worker.create({
+    connection: testEnv.nativeConnection,
+    taskQueue: 'test',
+    workflowsPath: require.resolve('../workflows'),
+    activities,
   });
+  await worker.runUntil(
+    testEnv.workflowClient.execute(processOrderWorkflow, {
+      workflowId: uuid(),
+      taskQueue: 'test',
+      args: [
+        {
+          orderProcessingMS: ms('3 days'),
+          sendDelayedEmailTimeoutMS: ms('1 day'),
+        },
+      ],
+    })
+  );
+  expect(emailSent).toBe(true);
+});
+```
+
+### Test arbitrary functions in Workflow context
+
+In case you need to test a function in your Workflow code that's not exported in [`workflowsPath`](https://typescript.temporal.io/api/interfaces/worker.workeroptions/#workflowspath), export it in a different path and register it with the Worker.
+
+`workflows/file-with-workflow-function-to-test.ts`
+
+```ts
+import * as wf from '@temporalio/workflow';
+import { someWorkflowToRunAsChild } from './some-workflow';
+
+export { someWorkflowToRunAsChild }; // Must be re-exported here for Worker registration
+
+export async function functionToTest() {
+  await wf.executeChild(someWorkflowToRunAsChild);
+  // Other test code
+}
+```
+
+`test.ts`
+
+```ts
+const worker = await Worker.create({
+  ...someOtherOptions,
+  connection: testEnv.nativeConnection,
+  workflowPath: require.resolve('./workflows/file-workflow-function-to-test'),
 });
+
+await worker.runUntil(
+  testEnv.workflowClient.execute(functionToTest, workflowOptions)
```

Comment:
Import `functionToTest`

Reply:
I'm debating whether to add this here, intentionally left out imports